src/ClientData/include/ClientData/CallstackData.h

```diff
@@ -83,9 +83,32 @@ class CallstackData {
   template <typename Action>
   void ForEachCallstackEventInTimeRangeDiscretized(uint64_t min_timestamp, uint64_t max_timestamp,
                                                    uint32_t resolution, Action&& action) const {
-    for (const auto& [tid, unused_events] : callstack_events_by_tid_) {
-      ForEachCallstackEventOfTidInTimeRangeDiscretized(tid, min_timestamp, max_timestamp,
-                                                       resolution, action);
+    auto get_next_callstack = [&](uint64_t timestamp) -> CallstackEvent {
+      const uint32_t kFakeId = -1;
+      CallstackEvent next_callstack{max_timestamp, kFakeId, kFakeId};
+      uint32_t current_pixel = GetPixelNumber(timestamp, resolution, min_timestamp, max_timestamp);
+      for (const auto& [unused_tid, events] : callstack_events_by_tid_) {
+        auto next_callstack_tid = events.lower_bound(timestamp);
+        if (next_callstack_tid != events.end() &&
+            next_callstack_tid->second.timestamp_ns() < next_callstack.timestamp_ns()) {
+          // If this callstack will be drawn in the current_pixel, we don't need to search for more
+          // of them, otherwise there could be a callstack in another thread_id that will be draw
+          // before, so we need to keep looking.
```

Comment:
```suggestion
          // If this callstack will be drawn in the current_pixel, we don't need to search for more
          // of them. Otherwise, there could be a callstack in another thread_id that will be drawn
          // before, so we need to keep looking.
```